doc/cprover-manual/function-contracts.md

```diff
@@ -0,0 +1,140 @@
+In our example, developers can use the built-in construct
+`__CPROVER_return_value`, which represents the return value of a function. As
+postconditions, one may list possible return values (in this
+case, either `SUCCESS` or `FAILURE`) as well as describe the main property of
+this function: if the function returns `SUCCESS`, then `*out` stores the
+result of `*a + *b`.
+We can also check that the value in `out` will be preserved in case of failure by using `__CPROVER_old`, which refers to the value of a given object in the pre-state of a function (see [History Varaibles]() Section for details). Thus, for the `sum` function, the set of postconditions are
+
+
+```c
+/* Postconditions */
+__CPROVER_ensures(__CPROVER_return_value == SUCCESS || __CPROVER_return_value == FAILURE)
+__CPROVER_ensures((__CPROVER_return_value == SUCCESS) ==> (*out == (*a + *b)))
+__CPROVER_ensures((__CPROVER_return_value == SUCCESS) ==> (*out == __CPROVER_old(*out)))
+```
+
+Finally, the _assigns_ clause allows developers to define a frame condition (see [Assigns Clauses](assigns-clause.md) Section for details).
+in general, be interpreted with either _writes_ or _modifies_ semantics, this
+design is based on the former. This means that memory not specified by the
+assigns clause must not be written within the given function scope, even if the
+value(s) therein are not modified. In our example, since we only expect
+that the value pointed by `out` may be modified, we annotate the function using
+`__CPROVER_assigns(*out)`.
+
+```c
+/* Writable Set */
+__CPROVER_assigns(*out)
+```
+
+Here is the whole function with its contract.
+
+```c
+int sum(uint32_t* a, uint32_t* b, uint32_t* out)
+/* Precondition */
+__CPROVER_requires(__CPROVER_is_fresh(a, sizeof(*a)))
+__CPROVER_requires(__CPROVER_is_fresh(b, sizeof(*b)))
+__CPROVER_requires(__CPROVER_is_fresh(out, sizeof(*out)))
+/* Postconditions */
+__CPROVER_ensures(__CPROVER_return_value == SUCCESS || __CPROVER_return_value == FAILURE)
+__CPROVER_ensures((__CPROVER_return_value == SUCCESS) ==> (*out == (*a + *b)))
+__CPROVER_ensures((__CPROVER_return_value == SUCCESS) ==> (*out == __CPROVER_old(*out)))
+/* Writable Set */
+__CPROVER_assigns(*out)
+{
+    const uint64_t result = ((uint64_t) *a) + ((uint64_t) *b);
+    if (result > UINT32_MAX) return FAILURE;
+    *out = (uint32_t) result;
+    return SUCCESS;
+}
+```
+
+First, we have to prove the function contract is correct.
+
+```shell
+   goto-cc -o main.goto *.c
+   goto-instrument --enforce-contract sum main.goto main-checking-contracts.goto
+   cbmc main-checking-contracts.goto
+```
+
+The first command just compiles the GOTO program as usual, the second command instruments the code to check the function contract, and the third one runs CBMC to do the checking.
+
+Now that we proved the function contract is correct, we can use the function contract in place of the function implementation wherever the function is called.
+
+```shell
+   goto-cc -o main.goto *.c
```

Comment:
There is no `main` function. Should `foo` above be `main`?

Reply:
Update it.